Replace ECDSA and Schnorr with Ed25519

somniumism opened the pull request.

somniumism commented on src/key/key.ts:
```diff
@@ -23,10 +20,17 @@ export const generatePrivateKey = (): string => {
-        .encode("hex")
-        .slice(2);
+    var enc = new TextEncoder();
+    return sodium.crypto_sign_ed25519_sk_to_pk(enc.encode(priv)).toString();
```
I'm trying to make`getPublicFromPrivate()` function. To get the public key in `sodiumoxide`, the following process is being used:
```rust
    pub fn public_key(&self) -> PublicKey {
        let mut pk = PublicKey([0u8; PUBLICKEYBYTES]);
        unsafe {
            ffi::crypto_sign_ed25519_sk_to_pk(pk.0.as_mut_ptr(), self.0.as_ptr());
        }
        pk
    }
```
However, `libsodium-wrappers` do not seem to provide the `crypto_sign_ed25519_sk_to_pk()` function. In [this file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/libsodium-wrappers/index.d.ts), we can check all the constant types and functions provided by `libsodium-wrappers`. But `crypto_sign_ed25519_sk_to_pk` is not there.

seonpyoKim replied:
Here is the wrapping symbol https://github.com/jedisct1/libsodium.js/blob/master/wrapper/symbols/crypto_sign_ed25519_sk_to_pk.json
The linked file is in Dfinity's repository.

somniumism replied:
I found it too, so I modified it. However, it still doesn't work properly. Informations about errors are here:
```
UnhandledPromiseRejectionWarning: TypeError: sodium.crypto_sign_ed25519_sk_to_pk is not a function
UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict`
[DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

seonpyoKim replied:
I read it but `libsodium-wrappers` does not expose the function. Let's use sumo version instead. It provides the function.

HoOngEe commented on src/key/eddsa.ts:
```diff
@@ -0,0 +1,74 @@
+import * as _ from "lodash";
```
Could you rename the file name "ed25519.ts"?

seonpyoKim commented on src/key/eddsa.ts:
```diff
@@ -0,0 +1,74 @@
+import * as _ from "lodash";
```
Could you rename the file name "ed25519.ts"?

somniumism replied:
I modified it.

joojis commented:
# [Codecov](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=h1) Report
> Merging [#1](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=desc) into [master](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/commit/1d30b6d97964cd11a87ab4d020d92dd397ffb4fd&el=desc) will **increase** coverage by `0.73%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1/graphs/tree.svg?width=650&height=150&src=pr&token=YbueYclDU1)](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master       #1      +/-   ##
==========================================
+ Coverage   91.75%   92.49%   +0.73%     
==========================================
  Files          16       15       -1     
  Lines         983      853     -130     
  Branches      216      180      -36     
==========================================
- Hits          902      789     -113     
+ Misses         81       64      -17     
```


| [Impacted Files](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/utility.ts](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1/diff?src=pr&el=tree#diff-c3JjL3V0aWxpdHkudHM=) | `76.92% <0.00%> (-23.08%)` | :arrow_down: |
| [src/key/ed25519.ts](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1/diff?src=pr&el=tree#diff-c3JjL2tleS9lZDI1NTE5LnRz) | `44.44% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=footer). Last update [1d30b6d...e7306e8](https://codecov.io/gh/CodeChain-io/foundry-primitives-js/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

seonpyoKim commented on src/key/key.ts:
```diff
@@ -1,32 +1,35 @@
- */
-const secp256k1 = new EC("secp256k1");
+export class Key {
+    public sodium: any;
```
We don't have to expose it to public

seonpyoKim commented on src/key/ed25519.ts:
```diff
@@ -0,0 +1,56 @@
+    };
+}
+
+/// test
```
You'd better to remove this test code or you can add it to test directory if needed.

seonpyoKim commented on test/key.test.ts:
```diff
@@ -1,100 +1,40 @@
-    verifyEcdsa,
-    verifySchnorr
+    Key,
+    // recoverEcdsa,
```
Remove them instead of making them comments

seonpyoKim commented on test/key.test.ts:
```diff
@@ -1,100 +1,40 @@
 });
 
-test("sign & verify ECDSA", () => {
+test("sign & verify EdDSA", async () => {
```
Let's use more specific name for the signature scheme, "Ed25519"

seonpyoKim replied:
I recommend you to generate testing oracles from other implementation like the previous.

somniumism commented:
I added tests that verify the result form my implementation like testing oracles. I used [pysodium](https://pypi.org/project/pysodium/) to make examples. @HoOngEe Could you check it?

somniumism commented:
Summary: I implemented three classes called `Key`, `Signature` and `Crypto`. The last class includes the former two classes since an async initialization process is needed to build functions in the `libsodium` crypto library. If this implementation is applied, all functions related to crypto and key that are being used by the foundry-keystone-js, as well as the foundry e2e test, should be changed. I want to know your opinion.

#### Introduction

I'm using the crypto library `libsodium` to replace ECDSA with Ed25519. To build functions in this library, we need an async initialization process called `.ready` for an object of `libsodium`. If there is no process, an error occurs indicating that the function in `libsodium` is not a function. That is,  for any function or structure that uses a function of libsodium, initializing an object is necessary. The process is as follows:
```typescript
/// Define the object 'sodium'
const sodium = require("libsodium-wrappers-sumo");
/// Initialize the object
await sodium.ready;
// Usage
const sk = sodium.generatePrivateKey();
```

To properly apply this initialization process, I discussed it with @HoOngEe  and @majecty . We judged that the process of initializing every time for every function is inefficient and may lead to a drop in performance. Thus, I implemented a classes `Key`, `Signature`, and `Crypto`, where the last class includes the first two classes mentioned. To build crypto functions, it is only necessary to call the initialization function within the class `Crypto` once an object is defined. The new process is as follows:
```typescript
// Define and initialize the object of class 'Crypto' that included the init process of 'sodium'.
const crypto = await Crypto.init();
// Usage
const sk = crypto.key.generatePrivateKey();
```

---

#### Implementation

```typescript
export class Crypto {
    public static async init() {
        const sodium = require("libsodium-wrappers-sumo");
        await sodium.ready;
        const crypto = new Crypto(sodium);

        return crypto;
    }

    public key: Key;
    public signature: Signature;
    private constructor(sodium: any) {
        this.key = new Key(sodium);
        this.signature = new Signature(sodium);
    }
}

class Key {
    private sodium: any;
    constructor(sodium: any) {
        this.sodium = sodium;
    }

    public generatePrivateKey(): string {
        // ...
    }

    public getPublicFromPrivate(privateKey: string): string {
         // ...
    }
}

class Signature {
    private sodium: any;
    constructor(sodium) {
        this.sodium = sodium;
    }
    public signEd25519(message, priv): string {
         // ...
    }
    public verifyEd25519 = (signature, message, pub): boolean => {
         // ...
        );
    };
}
```

---

#### Example

```typescript
/// no class(origin)
const sodium = require("libsodium-wrappers-sumo");
await sodium.ready;
const sk = sodium.generatePrivateKey();
const msg =
        "0000000000000000000000000000000000000000000000000000000000000000";
const sign = sodium.signEd25519(msg, sk);

/// only two classes 'Key' and 'Signature'
const key = new Key();
await key.init();
const sk = key.generatePrivateKey();
const msg =
        "0000000000000000000000000000000000000000000000000000000000000000";

const signature = new Signature();
await signature.init();
const sign = signature.signEd25519(msg, sk);


/// two classes 'Key' and 'Signature' with class 'Crypto'(the present)
const crypto = await Crypto.init();
const sk = crypto.key.generatePrivateKey();
const msg =
        "0000000000000000000000000000000000000000000000000000000000000000";
const sign = crypto.signature.signEd25519(msg, sk);
```

---

#### Conclusion

Functions related to key and crypto in the foundry-primitives-js is used in the foundry-keystone-js and foundry e2e tests. If this method is applied, the usage of all functions currently in use should change accordingly. Of course, it will not have a significant impact on the overall system. I want to know what you think about this implementation. Could you check this please?

HoOngEe commented:
It is off the point but IMO the private constructor should be empty: key and signature should be initialized in the `Crypto::init()`.

seonpyoKim commented on src/utility.ts:
```diff
@@ -19,13 +19,20 @@ export const toHex = (buffer: Buffer): string => {
+export const getAccountIdFromPrivate = async (
+    priv: string
+): Promise<string> => {
+    const crypto = await Crypto.init();
```
It is weird to create Crypto class every function call, we have two choices:
1. Move this function into the `Crypto` class.
2. Parameterize the `Crypto` context.

somniumism replied:
I'll choose 2. : )

seonpyoKim commented on src/key/crypto.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+}
+
+class Key {
```
We don't have to write all the classes in this file. What do you think of this change? @majecty

majecty replied:
I agree with you. Let's split the classes.

somniumism replied:
Can I combine class `Crypto` and class `Key` into key.ts file, and move class `Signature` to ed25519.ts file? It means two files, key.ts and ed25519.ts. Not three files, right?

somniumism replied:
I think it's better to separate into three files, `key.ts`, `ed25519.ts` and `crypto.ts`. I modified it.

sgkim126 commented:
@somniumism It seems that libsodium-wrappers-sumo is hard to use correctly, would you explain why it needs initialization?

somniumism commented:
@sgkim126 Actually, I don't understand why it needs initialization process `.ready` and internal process of `.ready`, but [libsodium-wrappers-sumo](https://www.npmjs.com/package/libsodium-wrappers-sumo) introduces the following:
> Load the sodium-wrappers module. The returned object contains a `.ready` property: a promise that must be resolve before the sodium functions can be used.

If the promise isn't resolved(the function of libsodium is not in the ready process), a TypeError occurs:

```typescript
const sodium = require("libsodium-wrappers-sumo");

console.log(toHex(sodium.crypto_sign_keypair().privateKey));
```
```
>> TypeError: sodium.crypto_sign_keypair is not a function
```
@HoOngEe I think my explanation is lacking, could you explain more?

HoOngEe commented:
@sgkim126 https://github.com/jedisct1/libsodium.js/issues/105 This may be the answer. However, I couldn't fully understand what exactly happens behind the `require("sodium-wappers-sumo")`.

sgkim126 commented:
It seems that it's to provide wasm implementation. I think there are three options.
1. Do not support web browsers. (wasm is an optimization for a web browser)
2. Find a pure javascript implementation.
3. Implement your design.

The first is not possible since the web is one of our primal goals.
I think your design is good if we cannot find a pure javascript implementation, or the found one is slow on a browser.

HoOngEe commented:
@sgkim126 [tweetnacl](https://github.com/dchest/tweetnacl-js#signatures) is a pure javascript implementation. If performance is not a concern then It seems fine.
I think the intention of the first choice can only be reached if omitting `await sodium.ready` deactivates wasm support only but working properly in other runtimes. Am I right?

sgkim126 commented:
> @sgkim126 [tweetnacl](https://github.com/dchest/tweetnacl-js#signatures) is a pure javascript implementation. If performance is not a concern then It seems fine.

@somniumism How about it?

> I think the intention of the first choice can only be reached if omitting `await sodium.ready` deactivates wasm support only but working properly in other runtimes. Am I right?

Yes. I thought that.

somniumism commented:
@sgkim126 The library is the result of the discussion with @HoOngEe . I'll replace ECDSA with Ed25519 using TweetNaCl. : ) Thank you.

HoOngEe commented:
Because we failed to find a way to deactivate wasm support selectively, pure javascript is the next best.

seonpyoKim commented on src/key/ed25519.ts:
```diff
@@ -0,0 +1,39 @@
+};
+
+/**
+ * Checks if the signature from signEd25519 is correct.
```
I think "valid" is correct rather than "correct" here. Let's modify the comments through this chance.

somniumism commented:
I replaced ECDSA and Schnorr with Ed25519 using crypto library `TweetNaCl`.
